### PR TITLE
feat: add TXDA vendor backend with op implementations and platform support

### PR DESCRIPTION
**Pull request by tsingmicro-public-e**
- Add TXDA (TsingMicro) vendor backend with implementations for silu_and_mul, rms_norm, gemma_rms_norm, rotary/mrotary_embedding, topk, fused_moe, and FLA ops (chunk_gated_delta_rule, fused_recurrent_gated_delta_rule)

- Add platform patches: dist_init (tccl backend + GroupCoordinator), fused_moe (moe_align_block_size/vllm_ops/unquant), model_runner (FL timer), per_rank_log redirection, platform_stubs

- Add tsingmicro.yaml platform config (prefer: flagos)

- Add torch-based fallback device detection when FlagGems DeviceDetector is unavailable

- Add TXDA device type mapping (tsingmicro → tccl) in platform and dispatch config

- Add get_compile_backend returning 'eager' for non-CUDA platforms

- Guard FlagGems import, suppress PyNccl when FlagCX is active, add platform activation fallback for txda

- Update examples for TXDA multi-node, concurrent, and offline inference scenarios

### CODE REVIEW

File: sglang_fl/__init__.py [hunk 1]
```diff
@@ -682,7 +694,17 @@ def activate_platform() -> str | None:
...
+        # Fallback: activate if txda is available via torch (e.g. when FlagGems
+        # DeviceDetector lacks txda support).
+        import torch as _torch
+
+        try:
+            import torch_txda  # noqa: F401
+        except ImportError:
+            pass
+        if hasattr(_torch, "txda") and _torch.txda.is_available():
+            return "sglang_fl.platform:PlatformFL"
```
Comment (huangxu0914) [inline sglang_fl/__init__.py]: The current version of FlagGems corresponding to sglang-plugin-FL is 5.3.0, which already supports tsingmicro vendors. If an exception occurs here, it indicates that there is a problem with FlagGems. Why use torch to forcefully load the backend?

File: sglang_fl/platform.py [hunk 2]
```diff
@@ -59,13 +60,29 @@ class PlatformFL(SRTPlatform):
...
-        self._vendor_name: str = detector.vendor_name  # "nvidia", "ascend", ...
-        self._device_type: str = detector.name  # "cuda", "npu", ...
-        self._dispatch_key: str = detector.dispatch_key  # "CUDA", "NPU", ...
-        self._device_count: int = detector.device_count
+        try:
+            # Core device identity from FlagGems
+            self._vendor_name: str = detector.vendor_name  # "nvidia", "ascend", ...
+            self._device_type: str = detector.name  # "cuda", "npu", ...
+            self._dispatch_key: str = detector.dispatch_key  # "CUDA", "NPU", ...
+            self._device_count: int = detector.device_count
```
Comment (huangxu0914) [inline sglang_fl/platform.py]: Should detector be initialized here?

File: sglang_fl/platform.py [hunk 3]
```diff
@@ -104,6 +121,20 @@ def _resolve_dist_backend(self) -> str:
...
+
+        if hasattr(_torch, "txda") and _torch.txda.is_available():
+            return ("txda", "txda", "TXDA", _torch.txda.device_count())
+        if hasattr(_torch, "npu") and _torch.npu.is_available():
+            return ("ascend", "npu", "NPU", _torch.npu.device_count())
+        if hasattr(_torch, "musa") and _torch.musa.is_available():
+            return ("mthreads", "musa", "MUSA", _torch.musa.device_count())
+        return ("nvidia", "cuda", "CUDA", _torch.cuda.device_count())
+
+
```
Comment (huangxu0914) [inline sglang_fl/platform.py]: The current device detection of sglang-plugin-FL is implemented through FlagGems. The torch detection here should be repeated.

File: sglang_fl/__init__.py [hunk 4]
```diff
@@ -466,6 +470,14 @@ def _init_hook(original_fn, self, *args, **kwargs):
...
                 logger.warning(f"CommunicatorFL creation failed: {e}")
                 self.fl_communicator = None
+            # Suppress PyNccl when FlagCX is active to avoid conflicts
+            if (
+                self.fl_communicator
+                and self.fl_communicator._flagcx_comm
+                and hasattr(self, "pynccl_comm")
+                and self.pynccl_comm is not None
+            ):
+                self.pynccl_comm.disabled = True
```
Comment (huangxu0914) [inline sglang_fl/__init__.py]: In a multi machine and multi rank, if CommunicatorFL is successfully loaded, pynccl has already been closed. If the load of CommunicatorFL fails, pynccl should not be closed. Is this modification because the tsingmicro platform needs to close pynccl?

File: sglang_fl/dispatch/config/utils.py [hunk 5]
```diff
@@ -25,6 +25,7 @@
 from pathlib import Path
 from typing import Any, Optional
 
+# import torch_txda
```
Comment (huangxu0914) [inline sglang_fl/dispatch/config/utils.py]: Can it be directly deleted?

File: sglang_fl/platform.py [hunk 6]
```diff
@@ -304,9 +315,14 @@ def init_backend(self) -> None:
...
-            f"sglang_fl.dispatch.backends.vendor.{self._vendor_name}.register_platform"
-        )
+        if self._vendor_name == "tsingmicro":
+            vendor_module = (
+                f"sglang_fl.dispatch.backends.vendor.txda.register_platform"
+            )
+        else:
+            vendor_module = (
+                f"sglang_fl.dispatch.backends.vendor.{self._vendor_name}.register_platform"
+            )
```
Comment (huangxu0914) [inline sglang_fl/platform.py]: Would it be more reasonable to rename sglang_fl.dispatch.backends.vendor.txda to sglang_fl.dispatch.backends.vendor.tsingmicro?

File: sglang_fl/__init__.py [hunk 7]
```diff
@@ -443,7 +443,10 @@ def _apply_vendor_patches() -> None:
         logger.warning("vendor patch skipped: DeviceDetector failed (%s)", e)
         return
 
-    module = f"sglang_fl.dispatch.backends.vendor.{vendor}.patch"
+    if vendor == "tsingmicro":
+        module = f"sglang_fl.dispatch.backends.vendor.txda.patch"
+    else:
+        module = f"sglang_fl.dispatch.backends.vendor.{vendor}.patch"
```
Comment (huangxu0914) [inline sglang_fl/__init__.py]: Would it be more reasonable to rename sglang_fl.dispatch.backends.vendor.txda to sglang_fl.dispatch.backends.vendor.tsingmicro?

File: examples/qwen3_6_35b_a3b_concurrent.py [hunk 8]
```diff
@@ -64,6 +73,26 @@
         "trust_remote_code": True,
         "disable_radix_cache": True,
     }
+elif _is_txda:
+    print("inference use txda")
```
Comment (yixiaodapeng) [inline examples/qwen3_6_35b_a3b_concurrent.py]: use logger please

File: examples/qwen3_6_35b_a3b_concurrent.py [hunk 9]
```diff
@@ -175,7 +204,7 @@ def _make_engine():
     return Engine(
         model_path=MODEL_PATH,
         tp_size=TP_SIZE,
-        mem_fraction_static=0.85,
+        # mem_fraction_static=0.85,
```
Comment (yixiaodapeng) [inline examples/qwen3_6_35b_a3b_concurrent.py]: remove pls

File: sglang_fl/dispatch/backends/vendor/txda/impl/normalization.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    out = torch.nn.functional.rms_norm(
+        x, (x.shape[-1],), obj.weight.data, obj.variance_epsilon
+    )
+    if residual is not None:
+        return out, residual
+    return out
+
+
+def gemma_rms_norm_txda(
+    obj,
```
Comment (yixiaodapeng) [inline sglang_fl/dispatch/backends/vendor/txda/impl/normalization.py]: why put  ref impl here？

File: AGENTS.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
+# AGENTS.md — sglang-plugin-FL
```
Comment (yixiaodapeng) [inline AGENTS.md]: remove md files pls

File: CLAUDE.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
+# CLAUDE.md
```
Comment (yixiaodapeng) [inline CLAUDE.md]: remove md files pls

File: examples/qwen3_6_27b_concurrent.py [hunk 13]
```diff
@@ -44,10 +50,15 @@
...
+if _is_txda:
+    os.environ.setdefault("SGLANG_FL_TIMER_ENABLE", "1")
+    os.environ.setdefault("SGLANG_REQ_WAITING_TIMEOUT", "-1")
+    os.environ.setdefault("SGLANG_REQ_RUNNING_TIMEOUT", "-1")
+
 # ─── Configuration ────────────────────────────────────────────────────────────
 
 MODEL_PATH = os.environ.get("MODEL_PATH", "/models/Qwen3.6-27B")
-TP_SIZE = int(os.environ.get("TP_SIZE", "4" if _is_npu else "1"))
+TP_SIZE = int(os.environ.get("TP_SIZE", "8" if _is_txda else ("4" if _is_npu else "1")))
```
Comment (ceci3) [inline examples/qwen3_6_27b_concurrent.py]: why example need different tp_size in same model?

File: examples/qwen3_6_27b_concurrent.py [hunk 14]
```diff
@@ -30,12 +30,18 @@
 import time
 from pathlib import Path
 
+try:
+    import torch_txda  # noqa: F401
+    from torch_txda import transfer_to_txda
+except ImportError:
+    pass
```
Comment (cyber-pioneer) [inline examples/qwen3_6_27b_concurrent.py]: this is not txda-relative file, should not import by default.

File: examples/qwen3_6_27b_concurrent.py [hunk 15]
```diff
@@ -175,7 +205,7 @@ def _make_engine():
     return Engine(
         model_path=MODEL_PATH,
         tp_size=TP_SIZE,
-        mem_fraction_static=0.85,
+        mem_fraction_static=0.6 if _is_txda else 0.85,
```
Comment (ceci3) [inline examples/qwen3_6_27b_concurrent.py:201]: why txda only need 0.6?

File: sglang_fl/dispatch/backends/vendor/tsingmicro/impl/fused_moe.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+    output = fused_experts_impl(
+        hidden_states=hidden_states,
+        w1=w1,
+        w2=w2,
+        topk_weights=topk_weights,
+        topk_ids=topk_ids,
+    )
+
+    return StandardCombineInput(hidden_states=output)
```
Comment (ceci3) [inline sglang_fl/dispatch/backends/vendor/tsingmicro/impl/fused_moe.py:40]: if txda use flaggems, why not dispatch to flaggems backends?

Comment (tsingmicro-public-e) [reply sglang_fl/dispatch/backends/vendor/tsingmicro/impl/fused_moe.py:40]: 使用 flaggems topk 算子有精度问题， 这个地方才这样适配的。

File: examples/qwen3_6_27b_concurrent.py (hunk 15 above)
Comment (tsingmicro-public-e) [reply examples/qwen3_6_27b_concurrent.py:201]: 使用 0.85 我们当前会OOM

File: examples/qwen3_6_27b_concurrent.py (hunk 13 above)
Comment (tsingmicro-public-e) [reply examples/qwen3_6_27b_concurrent.py]: 已经修改

File: examples/qwen3_6_27b_concurrent.py (hunk 14 above)
Comment (tsingmicro-public-e) [reply examples/qwen3_6_27b_concurrent.py]: 因为torch 没有注册txda 设备， _is_txda = hasattr(torch, "txda") and torch.txda.is_available() 这行代码必须依赖上面代码。

File: examples/qwen3_6_35b_a3b_concurrent.py (hunk 8 above)
Comment (tsingmicro-public-e) [reply examples/qwen3_6_35b_a3b_concurrent.py]: 已经删除

**Review by physics31415926 — APPROVED** [review]

File: sglang_fl/__init__.py [hunk 17]
```diff
@@ -489,14 +489,6 @@ def _init_hook(original_fn, self, *args, **kwargs):
...
                     ranks=self.ranks,
                 )
-                # Suppress PyNccl when FlagCX is active to avoid conflicts
-                if (
-                    self.fl_communicator
-                    and self.fl_communicator._flagcx_comm
-                    and hasattr(self, "pynccl_comm")
-                    and self.pynccl_comm is not None
-                ):
-                    self.pynccl_comm.disabled = True
```
Comment (huangxu0914) [inline sglang_fl/__init__.py:499]: please ensure that pynccl is disabled when flagcx is enabled.

File: sglang_fl/dispatch/backends/vendor/tsingmicro/impl/fused_moe.py (hunk 16 above)
Comment (yixiaodapeng) [reply sglang_fl/dispatch/backends/vendor/tsingmicro/impl/fused_moe.py:40]: if so，use blacklist to dispatch to ref. we are gonna give feedback to flaggems team internally

File: sglang_fl/__init__.py (hunk 17 above)
Comment (tsingmicro-public-e) [reply sglang_fl/__init__.py:499]: 已经fix

**Review by ceci3 — APPROVED** [review]

File: sglang_fl/dispatch/backends/vendor/tsingmicro/impl/activation.py [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+from __future__ import annotations
+
+import torch
+import torch.nn.functional as F
+
+
+def silu_and_mul_txda(obj, x: torch.Tensor) -> torch.Tensor:
+    d = x.shape[-1] // 2
+    x1, x2 = x[..., :d], x[..., d:]
+    return F.silu(x1) * x2
```
Comment (huangxu0914) [inline sglang_fl/dispatch/backends/vendor/tsingmicro/impl/activation.py]: Please remove the reference implementation.

**Review by yixiaodapeng — APPROVED** [review]
